src/ColumnMetricsMixin.js

```diff
@@ -51,41 +47,63 @@ module.exports = {
         columnMetrics.columns = this.adjustColumnWidths(columnMetrics.columns);
         this.setState(columnMetrics);
       } else {
-        var nextColumns = this.adjustColumnWidths(nextProps);
-        this.setState({columns: nextColumns});
+        var nextColumnStats = this.adjustColumnWidths(nextProps);
+        this.setState({columnMetrics: nextColumnStats});
       }
     }
   },
 
-  adjustColumnWidths(columns: ColumnMetricsType){
+  adjustColumnWidths(columnMetrics: ColumnMetricsType){
     var index = {};
-    this.state.columns.columns.forEach((c) => {
+    this.state.columnMetrics.columns.forEach((c) => {
       index[c.key] = {width: c.width, left: c.left};
     });
-    var nextColumns = Object.assign(this.state.columns, {
-      columns: columns.columns.map((c) => Object.assign(c, index[c.key]))
+    var nextColumns = Object.assign(this.state.columnMetrics, {
+      columns: columnMetrics.columns.map((c) => Object.assign(c, index[c.key]))
     });
     return nextColumns;
   },
 
-  getColumnMetricsType(props: ColumnMetricsType, initial: ?number): { columns: ColumnMetricsType; gridWidth: number } {
+  getColumnMetricsType(metrics: ColumnMetricsType, initial: ?number): { columns: ColumnMetricsType; gridWidth: number } {
     var totalWidth = initial ? initial : this.DOMMetrics.gridWidth();
-    return {
-      columns: ColumnMetrics.calculate({
-        columns: props.columns,
-        totalWidth: totalWidth,
-        minColumnWidth: props.minColumnWidth
-      }),
-      gridWidth: totalWidth
+    var currentMetrics = {
+      columns: metrics.columns,
+      totalWidth: totalWidth,
+      minColumnWidth: metrics.minColumnWidth
     };
+    var updatedMetrics
+    //if state has not yet been set or else if total width has changed then call recalculate
+    if(!this.state || (this.state && this.state.totalWidth !== totalWidth)){
+        updatedMetrics = ColumnMetrics.recalculate(currentMetrics);
+    } else{
+      updatedMetrics = currentMetrics;
+    }
+    return updatedMetrics;
+  },
+
+  getColumn(columns, idx) {
+    if(Array.isArray(columns)){
+      return columns[idx];
+    }else if (typeof Immutable !== 'undefined') {
+      return columns.get(idx);
+    }
+  },
+
+  getColumnCount() {
+    var columns = this.state.columnMetrics.columns;
+    if(Array.isArray(columns)){
```

Comment:
wonder if we should ship a flag for this to avoid the type checks?
either a high level state to indicate my cols are immutable, or even a code flag?